### PR TITLE
[SOT] Mark dynamic dims by type annotations

### PR DESCRIPTION
**Pull request by SigureMo**
利用类型提示来标记动态维度，以保证动转静能够一次即可收敛，避免重复组网

1. 支持递归处理 typing，比如 `dataclass`、`Optional[T]`
2. Tensor 类型具有隐式动态维度 `(0,)`
3. 支持以 `Annotated[Tensor, DynamicDims((1, 2))]` 标记其余维度

示例如下

```python
from typing import Annotated
from dataclasses import dataclass

from fastdeploy.model_executor.graph_optimization.dynamic_dims_marker import (
    DynamicDims
)
def fn(x: Tensor):
    # x 具有隐式动态维度 (0,)
    ...

def fn(x: Annotated[Tensor, DynamicDims(1)]):
    # x 具有动态维度 (1,)
    ...

@dataclass
class ForwardMeta:
    a: Annotated[Tensor, DynamicDims((2, 3))]
    caches: list[Tensor]

def fn(x: ForwardMeta):
    # x.a 具有动态维度 (2, 3)
    # x.caches 中所有元素具有隐式动态维度 (0,)
    ...
```

另外在 warmup 阶段会根据是否发生打断、是否有重复构图来确定能否用更加快速的实现，直接使用编译好的 code 来执行，当无法使用时会报 warning

### CODE REVIEW

**Comment by gongshaotian** [comment]
如何理解 Tensor 类型具有隐式动态维度 (0,)，是一个规则吗

File: fastdeploy/model_executor/graph_optimization/dynamic_dims_marker.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+class DynamicDims:
+    def __init__(self, dims: int | tuple[int]):
+        self.dims = dims if isinstance(dims, tuple) else (dims,)
+
+    def __repr__(self):
+        return f"DynamicDims({self.dims})"
+
+
+class DynamicDimTypeResolver:
```
Comment (gongshaotian) [inline fastdeploy/model_executor/graph_optimization/dynamic_dims_marker.py:45]: 建议写个基类注释

File: fastdeploy/model_executor/graph_optimization/graph_optimization_backend.py [hunk 2]
```diff
@@ -14,14 +14,97 @@
...
+def apply_to_static_optimization(fn, backend: ToStaticBackend):
+    forward_fn = fn
+    forward_sig = inspect.signature(forward_fn)
+    forward_type_hints = get_type_hints(forward_fn)
+    static_forward_fn = sot.symbolic_translate(forward_fn, training=False, backend=backend)
+    unsafe_static_forward_fn = None
+    need_warmup = True
+
+    @functools.wraps(forward_fn)
+    def warmup_impl(self, *args, **kwargs):
```
Comment (gongshaotian) [inline fastdeploy/model_executor/graph_optimization/graph_optimization_backend.py:59]: 返回类型是什么

Comment (gongshaotian) [inline fastdeploy/model_executor/graph_optimization/graph_optimization_backend.py:59]: 返回类型是什么，tensor 吗

File: fastdeploy/model_executor/graph_optimization/graph_optimization_backend.py [hunk 3]
```diff
@@ -14,14 +14,97 @@
...
+            forward_fn.__name__,
+            forward_fn.__defaults__,
+            forward_fn.__closure__,
+        )
+        return result
+
+    @functools.wraps(forward_fn)
+    def static_forward(self, *args, **kwargs):
+        nonlocal need_warmup
+        is_warmup = in_warmup_mode() and need_warmup
```
Comment (gongshaotian) [inline fastdeploy/model_executor/graph_optimization/graph_optimization_backend.py:103]: in_warmup_mode 的作用是什么，dummy run 32次之后再开始构图？

**Comment by SigureMo** [comment]
> 如何理解 Tensor 类型具有隐式动态维度 (0,)，是一个规则吗

是一个规则，主要是基本上所有 Tensor batch 维度基本都是动态的，从标记的角度将会很麻烦，因此添加此默认规则（该隐式规则同 vLLM）

File: fastdeploy/model_executor/graph_optimization/graph_optimization_backend.py (hunk 3 above)
Comment (SigureMo) [reply fastdeploy/model_executor/graph_optimization/graph_optimization_backend.py:103]: 这里的 `in_warmup_mode` 会替换为 #2929 里的实现，目前两边开发是解耦的

构图是在一开始就进行的，warmup 阶段会确定两件事：

1. 是否有打断
2. 是否有重复组网

如果没有上述问题，则可以直接使用加速策略（直接使用 code 跑），否则使用默认策略

File: fastdeploy/model_executor/graph_optimization/dynamic_dims_marker.py (hunk 1 above)
Comment (SigureMo) [reply fastdeploy/model_executor/graph_optimization/dynamic_dims_marker.py:45]: 已添加

**Review by gongshaotian — DISMISSED** [review]
LGTM

File: fastdeploy/model_executor/models/ernie4_5_vl/ernie4_5_vl_moe.py [hunk 4]
```diff
@@ -406,7 +406,7 @@ def load_state_dict(self, state_dict):
     def forward(
         self,
         ids_remove_padding: paddle.Tensor,
-        image_features: paddle.Tensor,
+        image_features: Optional[paddle.Tensor],
```
Comment (yuanlehome) [inline fastdeploy/model_executor/models/ernie4_5_vl/ernie4_5_vl_moe.py:409]: 其他forward函数中也写成 Optional[paddle.Tensor] 吧

File: fastdeploy/model_executor/layers/attention/xpu_attn_backend.py [hunk 5]
```diff
@@ -62,20 +62,23 @@ class XPUAttentionMetadata(AttentionMetadata):
     block_tables: Optional[paddle.Tensor] = None
     rotary_embs: Optional[paddle.Tensor] = None
     attn_mask: Optional[paddle.Tensor] = None
-    encoder_block_shape_q: Optional[paddle.Tensor] = None
-    decoder_block_shape_q: Optional[paddle.Tensor] = None
+    encoder_block_shape_q: int = -1
+    decoder_block_shape_q: int = -1
```
Comment (yuanlehome) [inline fastdeploy/model_executor/layers/attention/xpu_attn_backend.py:66]: int ?

File: fastdeploy/model_executor/layers/attention/__init__.py [hunk 6]
```diff
@@ -32,4 +33,5 @@
     "FlashAttentionBackend",
     "IluvatarAttnBackend",
     "BlockAttentionBackend",
+    "Attention",
```
Comment (yuanlehome) [inline fastdeploy/model_executor/layers/attention/__init__.py:36]: 加上这个我记得是会有循环引用的？

File: fastdeploy/model_executor/graph_optimization/dynamic_dims_marker.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+    @abstractmethod
+    def type_match(self, tp) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def extract_inner_types(self, data, data_name, tp) -> list[tuple[Accessor[Any, Any], str, type[Any]]]:
+        raise NotImplementedError
+
+    def resolve(self, data, data_name, tp) -> None:
```
Comment (yuanlehome) [inline fastdeploy/model_executor/graph_optimization/dynamic_dims_marker.py]: 所有函数参数都需要添加类型注解

File: fastdeploy/model_executor/models/ernie4_5_vl/ernie4_5_vl_moe.py (hunk 4 above)
Comment (SigureMo) [reply fastdeploy/model_executor/models/ernie4_5_vl/ernie4_5_vl_moe.py:409]: 已修改

File: fastdeploy/model_executor/layers/attention/__init__.py (hunk 6 above)
Comment (SigureMo) [reply fastdeploy/model_executor/layers/attention/__init__.py:36]: 不会有的，除非是之前设计不合理导致的

File: fastdeploy/model_executor/graph_optimization/dynamic_dims_marker.py (hunk 7 above)
Comment (SigureMo) [reply fastdeploy/model_executor/graph_optimization/dynamic_dims_marker.py]: 已经添加

File: fastdeploy/engine/engine.py [hunk 8]
```diff
@@ -937,11 +937,11 @@ def _setting_environ_variables(self):
                 "SOT_LOG_LEVEL": os.getenv("SOT_LOG_LEVEL", default="0"),
                 "SOT_UNSAFE_CACHE_FASTPATH": os.getenv("SOT_UNSAFE_CACHE_FASTPATH", default="1"),
                 "SOT_ENABLE_0_SIZE_FALLBACK": os.getenv("SOT_ENABLE_0_SIZE_FALLBACK", default="0"),
+                "SOT_SPECIALIZED_DIM_NUMBERS": os.getenv("SOT_SPECIALIZED_DIM_NUMBERS", default="no"),
```
Comment (zyfncg) [inline fastdeploy/engine/engine.py:940]: 这个默认是 no ?

File: fastdeploy/model_executor/layers/attention/flash_attn_backend.py [hunk 9]
```diff
@@ -144,7 +147,7 @@ def get_kv_cache_shape(
             self.head_dim,
         )
 
-    def init_attention_metadata(self, forward_meta: ForwardMeta):
+    def init_attention_metadata(self, forward_meta: "ForwardMeta"):
```
Comment (zyfncg) [inline fastdeploy/model_executor/layers/attention/flash_attn_backend.py]: "ForwardMeta" 包字符串的作用是什么？

File: fastdeploy/engine/engine.py (hunk 8 above)
Comment (SigureMo) [reply fastdeploy/engine/engine.py:940]: 框架里默认特化 1，即 `"1"`，FD 里默认是不特化，是 `"no"`

File: fastdeploy/model_executor/layers/attention/flash_attn_backend.py (hunk 9 above)
Comment (SigureMo) [reply fastdeploy/model_executor/layers/attention/flash_attn_backend.py]: 常用于

- 解决循环引用
- 降低运行时解析开销

但这些文件都使用了 PEP 563，所以这里其实加不加 `"` 效果是一样的，为提高可读性这里恢复了下

**Review by yuanlehome — APPROVED** [review]
